pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/CursorIntegrationTest.java

```diff
@@ -0,0 +1,432 @@
+    }
+  }
+
+  @Test
+  public void testBadDelete() {
+    try {
+      ClusterTest.sendDeleteRequest(getBrokerDeleteQueryStoresApiUrl(getBrokerBaseApiUrl(), "dummy"), getHeaders());
+    } catch (IOException e) {
+      HttpErrorStatusException h = (HttpErrorStatusException) e.getCause();
+      Assert.assertEquals(h.getStatusCode(), 404);
+      Assert.assertTrue(h.getMessage().contains("Query results for dummy not found"));
+    }
+  }
+
+  @Test
+  public void testQueryWithEmptyResult()
+      throws Exception {
+    JsonNode pinotResponse = ClusterTest.postQuery(EMPTY_RESULT_QUERY,
+        getBrokerQueryApiUrl(getBrokerBaseApiUrl()) + getCursorQueryProperties(1000), getHeaders(),
+        getExtraQueryProperties());
+
+    // There should be no resultTable.
+    Assert.assertNull(pinotResponse.get("resultTable"));
+    // Total Rows in result set should be 0.
+    Assert.assertEquals(pinotResponse.get("numRowsResultSet").asInt(), 0);
+    // Rows in the current response should be 0
+    Assert.assertEquals(pinotResponse.get("numRows").asInt(), 0);
+    Assert.assertTrue(pinotResponse.get("exceptions").isEmpty());
+  }
+
+  @DataProvider(name = "InvalidOffsetQueryProvider")
+  public Object[][] invalidOffsetQueryProvider() {
+    return new Object[][]{{TEST_QUERY_ONE}, {EMPTY_RESULT_QUERY}};
+  }
+
+  @Test(dataProvider = "InvalidOffsetQueryProvider", expectedExceptions = IOException.class,
+      expectedExceptionsMessageRegExp = ".*Offset \\d+ should be lesser than totalRecords \\d+.*")
+  public void testGetInvalidOffset(String query)
+      throws Exception {
+    CursorResponse pinotPagingResponse;
+    pinotPagingResponse = JsonUtils.jsonNodeToObject(ClusterTest.postQuery(query,
+        getBrokerQueryApiUrl(getBrokerBaseApiUrl()) + getCursorQueryProperties(_resultSize), getHeaders(),
+        getExtraQueryProperties()), CursorResponseNative.class);
+    Assert.assertTrue(pinotPagingResponse.getExceptions().isEmpty());
+    ClusterTest.sendGetRequest(
+        getBrokerResponseApiUrl(getBrokerBaseApiUrl(), pinotPagingResponse.getRequestId()) + getCursorOffset(
+            pinotPagingResponse.getNumRowsResultSet() + 1), getHeaders());
+  }
+
+  @Test
+  public void testQueryWithRuntimeError()
+      throws Exception {
+    String queryWithFromMissing = "SELECT * mytable limit 100";
+    JsonNode pinotResponse;
+    pinotResponse = ClusterTest.postQuery(queryWithFromMissing,
+        getBrokerQueryApiUrl(getBrokerBaseApiUrl()) + getCursorQueryProperties(_resultSize), getHeaders(),
+        getExtraQueryProperties());
+    Assert.assertFalse(pinotResponse.get("exceptions").isEmpty());
+    JsonNode exception = pinotResponse.get("exceptions").get(0);
+    Assert.assertTrue(exception.get("message").asText().startsWith("QueryValidationError:"));
+    Assert.assertEquals(exception.get("errorCode").asInt(), 700);
+    Assert.assertTrue(pinotResponse.get("brokerId").asText().startsWith("Broker_"));
```

Comment:
Maybe we can also verify that the get all response stores API returns 0 results in this case?

Reply:
Oh, I meant something like this:

```
    List<CursorResponseNative> requestIds = JsonUtils.stringToObject(
        ClusterTest.sendGetRequest(getBrokerGetAllResponseStoresApiUrl(getBrokerBaseApiUrl()), getHeaders()),
        new TypeReference<>() {
        });

    Assert.assertEquals(requestIds.size(), 0);
```

But it was just an optional suggestion anyway.